Scripts/static-analyzer/Sources/Analyzer/SourceNavigator.swift

```diff
@@ -0,0 +1,160 @@
+        let fileManager = FileManager.default
+        let tempFile = fileManager.temporaryDirectory.appendingPathComponent("temp.swift")
+        var tempFileContent = try String(contentsOfFile: location.path)
+
+        if let moduleName, moduleName.starts(with: "_") == false, moduleName != appModuleName {
+            tempFileContent = tempFileContent.inserting("import \(moduleName)", atLine: 1)
+        }
+
+
+        let typeCheckCode = "let __injected_variable: \(typename)? = nil"
+        let typenameLine = location.line + 1
+        let typenameColumn = typeCheckCode.range(of: typename)!.upperBound.utf16Offset(in: typeCheckCode)
+        tempFileContent = tempFileContent.inserting(typeCheckCode, atLine: typenameLine)
+
+        try tempFileContent.write(to: tempFile, atomically: true, encoding: .utf8)
+        let typenameOffset = StringView(tempFileContent).byteOffset(forLine: Int64(typenameLine), bytePosition: Int64(typenameColumn))!
+
+        let newArgs = compilerArguments.map({ arg in
+            if arg == location.path {
+                return tempFile.path
+            } else {
+                return arg
+            }
+        })
+
+        let response = try Request.cursorInfo(file: tempFile.path, offset: typenameOffset, arguments: newArgs).send()
+        try response.ensureSourceKitSuccessfulReponse()
+
+        let kind: String = try response.get("key.kind")
+        guard kind == "source.lang.swift.ref.class" else {
+            print("\(typename) is \(kind)")
+            return false
+        }
+
+        let decl: String = try response.get("key.fully_annotated_decl")
+        let regex = /<ref\.class.*>(.+)<\/ref.class>/
+        guard let group = try regex.firstMatch(in: decl)?.1 else {
+            throw AnyError(message: "Can't parse superclass name from declaration: \(decl)")
+        }
+
+        let superTypename = String(group)
+        switch superTypename {
+        case superclass: return true
+        case "NSObject": return false
+        case "MainActor": return false
+        default:
+            let moduleName: String = try response.get("key.modulename")
+            return try isInheritence(subclass: superTypename, moduleName: moduleName, superclass: superclass, usedAt: location, compilerArguments: compilerArguments)
+        }
+    }
+
+}
+
+private extension String {
+    func inserting(_ string: String, atLine line: Int) -> String {
+        precondition(line > 0)
+
+        var allLines = [String]()
+        self.enumerateLines { line, _ in
+            allLines.append(line)
+        }
+        allLines.insert(string, at: line - 1)
+        return allLines.joined(separator: "\n")
+    }
+}
+
+extension IndexSymbolKind {
+    var isTypeDefinition: Bool {
+        switch self {
+        case .enum, .struct, .class, .protocol:
+            return true
```

Comment:
I was surprised not to find documentation for this type. I found [this post](https://cheekyghost.com/indexstore-for-swift) which links to [code that exhaustively enumerates the cases](https://github.com/CheekyGhost-Labs/IndexStore/blob/5f63b76bfa799b063cb348b55caabea2a3e9e75c/Sources/IndexStore/Public/SourceSymbol/SourceKind.swift#L119-L180).

I found `typealias` in there, would that count as a type definition for the purpose of this tool?

_Side note: also `concept` and I don't know what it is._ 🤔